src/tpm2tss_engine.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2020 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <openssl/ssl.h>
+#include <openssl/ui.h>
+#include <openssl/engine.h>
+#include <stdio.h>
+#include <string.h>
+#include "log.h"
+#include "tpm2tss_engine.h"
+
+static ENGINE *
+enftun_tpm2tss_engine_init()
+{
+    ENGINE *eng = NULL;
+    const char *eng_name = "tpm2tss";
+
+    /* Init the engine */
+    ENGINE_load_builtin_engines();
+
+    eng = ENGINE_by_id(eng_name);
+    if (!eng) {
+        fprintf(stderr, "%s ERR: Could not find engine\n", __func__);
+        goto out;
+    }
+
+    if (!ENGINE_init(eng)) {
+        fprintf(stderr, "%s ERR: Could not tpm2tss engine\n", __func__);
+        ENGINE_free(eng);
+        goto out;
+    }
+
+    if (!ENGINE_set_default(eng, ENGINE_METHOD_ALL)) {
+        fprintf(stderr, "%s ERR: Could not set tpm2tss engine as default\n", __func__);
+        ENGINE_finish(eng);
+        ENGINE_free(eng);
+        goto out;
+    }
+
+    out:
+    return eng;
+}
+
+int
+enftun_tpm2tss_ssl_ctx_set(const char *key_file, SSL_CTX* ctx)
+{
+    ENGINE *e;
+    EVP_PKEY * key = NULL;
+    UI_METHOD *meth = NULL;
+    int ret = 0;
+
+    e = enftun_tpm2tss_engine_init();
+    if (!e)
+    {
+        enftun_log_ssl_error("Failed to load engine tpm2tss");
+        ret = 1;
+        goto out;
+    }
+
+    meth = enftun_ssl_ui_passthrough();
```

Comment:
Let's be sure to test with a key with password.  `enftun` wasn't written to support any UI, so I'm not sure how the built-in OpenSSL UI will work.  The tpm2-tss-engine [supports](https://github.com/tpm2-software/tpm2-tss-engine/blob/62a60ef9aa08726afa72dcd5361c4089ff78a263/src/tpm2-tss-engine.c#L73) a password, so it should be possible to test.

I don't see a use-case for `enftun` with a password-protected key, but we should test that this code works.